### PR TITLE
Custom actions/block modules

### PR DESCRIPTION
**Pull request by emmamedeiros**
This pull request contains the control block and ast modules for custom-actions implementation, which include ast_block.h, ast_block.c, test_ast_block.c, control_block.h, control_block.c, test_control_block.c files. These outline and implement new, init, and free functions for each type of block in this pull request. A separate pull request for the other block modules will follow.

### CODE REVIEW

File: src/custom-actions/include/action_block.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+/* Forward declaration */
+typedef struct AST_block AST_block_t;
+
+/* An enumeration type for an action block */
+typedef enum action_type {
+    SET,
+    SAY,
+    MOVE, 
+    ADDSUBMULTDIV,
```
Comment (brentd313) [inline src/custom-actions/include/action_block.h:20]: Maybe make this something like ARITHMETIC

Comment (emmamedeiros) [reply src/custom-actions/include/action_block.h:20]: sounds good! I can change that

**Comment by emmamedeiros** [comment]
The ADDSUBMULTDIV action_type_t was changed to ARITHMETIC during the review, and the pull request was approved and merged into dev.

File: src/custom-actions/include/ast_block.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ * - A block. 
+ */
+AST_block_t* AST_block_new(block_t *block, block_type_t block_type);
+
+/* 
+ * Initializes an AST block. 
+ * 
+ * Parameters: 
+ * - AST block. Must point to already allocated memory. 
+ * - block, block_type: parameters that define the block
```
Comment (MaxineK36) [inline src/custom-actions/include/ast_block.h:55]: please put each parameter on its own line

File: src/custom-actions/include/branch_block.h [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ * - integer containing the number of conditional blocks
+ * - pointer to a list of conditional blocks
+ * - integer for the number of control blocks
+ * - pointer to a list of control blocks 
+ * - enum representing the conditional type 
+ * 
+ * Returns: 
+ * - A branch block. 
+ */  
+branch_block_t* branch_block_new(int num_conditionals, conditional_block_t**
```
Comment (MaxineK36) [inline src/custom-actions/include/branch_block.h:39]: please make the lines of parameters line up (see more info on long lines in the style guide [here](https://uchicago-cs.github.io/dev-guide/style_guide.html#maximum-line-length))

File: src/custom-actions/src/action_block.c [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        return NULL;    
+    }
+    
+    return action;
+}
+
+/* See action_block.h */
+AST_block_t* AST_action_block_new(action_enum_t action_type, int num_args, 
+                                  attribute_t** args)
+{
```
Comment (MaxineK36) [inline src/custom-actions/src/action_block.c:43]: This function header looks great! This is how all the functions should be aligned

File: src/custom-actions/src/action_block.c [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    AST_block_t *ast;
+    action_block_t *action;
+    int new_action;
+    block_type_t block_type = ACTION;
+
```
Comment (MaxineK36) [inline src/custom-actions/src/action_block.c:48]: This function could use some documentation (i.e. comments) explaining a little bit what's happening (i.e. action block is created, nested in a block and in an ast_block)

File: src/custom-actions/src/action_block.c [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    new_action = action_block_init(action, action_type, num_args, args);
+    if (new_action != SUCCESS) 
+    {
+        fprintf(stderr, "Error: Could not initialize action_block_t\n");
+        return NULL;    
+    }
+    
+    block_t* block = malloc(sizeof(block));
```
Comment (MaxineK36) [inline src/custom-actions/src/action_block.c:65]: should be `sizeof(block_t)`

File: src/custom-actions/include/branch_block.h [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+ * Returns: 
+ * - SUCCESS if success, FAILURE if error occurs
+ */  
+int branch_block_init(branch_block_t *branch, int num_conditionals, conditional_block_t**
+conditionals, conditional_type_t conditional_type, int num_controls,
+control_block_t** controls);
+
+/* 
+ * Frees a branch block. 
+ *
```
Comment (MaxineK36) [inline src/custom-actions/include/branch_block.h:80]: Does it free the nested blocks as well?

File: src/custom-actions/CMakeLists.txt [hunk 8]
```diff
@@ -1,4 +1,8 @@
 add_library(custom-actions
+            src/action_block.c
```
Comment (MaxineK36) [inline src/custom-actions/CMakeLists.txt:2]: I thought this PR only had ast_block.c and control_block.c

File: src/custom-actions/src/ast_block.c [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+#include <stdlib.h>
+#include <stdio.h>
+#include "game-state/item.h" 
+#include "ast_block.h"
+
+/* See ast_block.h */
+AST_block_t* AST_block_new(block_t *block, block_type_t block_type)
+{
+    AST_block_t *ast;
```
Comment (MaxineK36) [inline src/custom-actions/src/ast_block.c:16]: lines 16 and 19 can be condensed into one line

File: src/custom-actions/src/ast_block.c [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        return NULL;
+    }
+
+    return ast; 
+}
+
+/* See ast_block.h */
+int AST_block_init(AST_block_t *ast, block_t *block, block_type_t block_type)
+{
+    assert(ast != NULL);
```
Comment (MaxineK36) [inline src/custom-actions/src/ast_block.c:40]: did you want to also guarantee that the block isn't null?

File: src/custom-actions/src/ast_block.c [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    ast->block_type = block_type;
+    
+    return SUCCESS; 
+}
+
+/* See ast_block.h */
+int AST_block_free(AST_block_t *ast)
+{
+    assert(ast != NULL);
+
```
Comment (MaxineK36) [inline src/custom-actions/src/ast_block.c:52]: AST_block_free should also free the nested block

File: src/custom-actions/src/conditional_block.c [hunk 12]
```diff
@@ -0,0 +1,96 @@
+/*
```
Comment (MaxineK36) [inline src/custom-actions/src/conditional_block.c:1]: The PR indicates that this file shouldn't have been included

File: src/custom-actions/src/action_block.c [hunk 13]
```diff
@@ -0,0 +1,94 @@
+/*
```
Comment (MaxineK36) [inline src/custom-actions/src/action_block.c:1]: The PR indicates that this file shouldn't have been included

File: tests/custom-actions/CMakeLists.txt [hunk 14]
```diff
@@ -1,6 +1,10 @@
 set(TEST_EXE test-custom-actions)
 
 add_executable(${TEST_EXE}
+        test_action_block.c
```
Comment (MaxineK36) [inline tests/custom-actions/CMakeLists.txt:4]: I thought this PR only had ast and control

**Comment by emmamedeiros** [comment]
AST block and control block modules have been merged into dev, and thus I will be closing this pull request.

**Comment by MaxineK36** [comment]
Issue Score: ✔️++

Comments:
Great work! The explicit update about the `action_type_t` change was very useful.

**Pull request merged**